flang/lib/Lower/Bridge.cpp

```diff
@@ -4290,6 +4291,18 @@ class FirConverter : public Fortran::lower::AbstractConverter {
     assert(blockId == 0 && "invalid blockId");
     assert(activeConstructStack.empty() && "invalid construct stack state");
 
+    // Set target_cpu and target_features attributes to be passed through to the
+    // llvm.func operation during lowering.
+    const llvm::TargetMachine &targetMachine = bridge.getTargetMachine();
+    if (auto targetCPU = targetMachine.getTargetCPU(); !targetCPU.empty())
+      func->setAttr("target_cpu",
+                    mlir::StringAttr::get(func.getContext(), targetCPU));
+
+    if (auto targetFeatures = targetMachine.getTargetFeatureString();
+        !targetFeatures.empty())
+      func->setAttr("target_features", mlir::LLVM::TargetFeaturesAttr::get(
+                                           func.getContext(), targetFeatures));
```

Comment:
In FrontendAction we already have the TargetMachine, and we use a part of it in the bridge, so this change does not introduce any new dependencies.  Not that it would be wrong to do so, but it shows that we already use a part of the information from the target machine.  I think it is logical to take it further, and include _all_ information from TargetMachine into the lowered MLIR.

Reply:
@kiranchandramohan at this time, these function attributes would be added with the only purpose of being forwarded to the LLVM dialect and from there being translated into LLVM IR function attributes, so as you say it would be possible to delay their introduction to a later pass rather than doing it directly in lowering.

However, in my opinion, this is general information in the same category as the target triple, kind mapping or data layout that could be at some point needed earlier in the compilation flow, just like these other attributes are. I agree with @kparzysz that having this information available in the lowering bridge doesn't really introduce new dependencies, so I don't currently see any clear downsides. The benefit of this approach is that this information is available as soon as possible and we don't introduce intermediate MLIR representations where this information that is known in advance can't be accessed.

So, I think at the moment this is still my preferred approach, though I'm happy to change it if there are any other aspects I've missed or if the alternative seems to be generally preferred by the community.

Reply:
I _think_ this doesn't set the target info for function declarations (interfaces that are in a different source), which I  need. I'm happy to add that as an addition on top. l

Are we getting the target CPU and target features for every function? Or just the once? I'm sure it's not a "heavy" operaiton, but seems a bit unnecessary to do repeatedly. Should be the same for ALL functions in a module...

Reply:
@Leporacanthicus thanks for the feedback! I didn't think about testing function declarations, so I'll make some tests and let you know. If that doesn't work, then it might be a reason to go for the pass-based approach depending on how difficult that case would be to address in lowering.

About your second concern, I think it should be fine as it is right now. We can cache the target CPU and features strings somewhere in the bridge to avoid calling `TargetMachine::getTarget{CPU, FeatureString}` for each function, but these functions just return cached `StringRef`s, so I think it's probably not a problem.

Reply:
I am in favor of delaying this to something like codegen (or better, the target rewrite pass) based on Kiran point that functions may be generated during FIR to FIR passes and that this is a lot of noise.

Some target specific aspect are unavoidable in high level representation because it is needed for Fortran semantics (e.g., the data layout), but it does not mean we should start doing/adding a lot of target specific aspects in the IR before it is actually needed.

I would also prefer passing the target_cpu/target_features as strings to the target rewrite pass, the TargetMachine is not an easy object to come by, and makes testing dependent on the available targets in the LLVM build.

Reply:
Thank you all for sharing your suggestions and concerns, I can now understand better the problems with this approach so I'll work on following @jeanPerier's suggested alternative and move this logic from the lowering bridge to the target rewrite pass. This should also address the annotation of function declarations raised by @Leporacanthicus, which I tested to not be covered by this patch in its current form.

Reply:
The intent in adding all relevant information from target machine early on is to become independent of it, or in other words, the generated module is more self-contained.  We already get the target machine at this point, so we may extract all that we're going to need from it, until having it again becomes necessary (generation of LLVM IR?).

This isn't to say that we have to do it, it just explains my motive for suggesting it.

Reply:
> We already get the target machine at this point, so we may extract all that we're going to need from it, until having it again becomes necessary (generation of LLVM IR?).

The issue is that this implies that every func.func created after the bridge/not in this very spot would not have these attributes while they should. So this would push the requirements of such target attribute setting to every spot creating fun.func. We could use a centralized helper, but you would still need to thread the info to the place calling it, and you cannot easily enforce that it is being used (especially if some FIR external passes are used).

So, much safer/consistent to do it in a pass IMHO.

I understand your point about self-contained module, and I am not opposed adding these attributes to the module op until we are ready to add them on the func.func (that is, before codegen when we should not emit new func.func), and to retrieve them from here instead of the driver.

Reply:
> I understand your point about self-contained module, and I am not opposed adding these attributes to the module op until we are ready to add them on the func.func (that is, before codegen when we should not emit new func.func), and to retrieve them from here instead of the driver.

I think this is a good compromise. Since these attributes will have the same values for all functions, we can attach this information to the module, making it available as early as possible, but without having to deal with explicitly adding them to each function that is created. Then, the target rewrite pass only has to forward these module attributes to each function.